dali/kernels/common/block_setup.h

```diff
@@ -0,0 +1,263 @@
+  return ret;
+}
+
+template <int skip, int n>
+DALI_HOST_DEV
+std::enable_if_t<(skip < 0), TensorShape<n>> skip_dim(const TensorShape<n> &shape) {
+  return shape;
+}
+
+template <int skip, int n>
+DALI_HOST_DEV
+std::enable_if_t<(skip >= 0), TensorShape<n-1>> skip_dim(const TensorShape<n> &shape) {
+  static_assert(skip < n, "The dimension to be skipped must not exceed input ndim");
+  return shape_cat(shape.template first<skip>(), shape.template last<n-skip-1>());
+}
+
+/// @brief A utility for calculating block layout for GPU kernels
+/// @tparam _ndim         number dimensions to take into account while calculating the layout
+/// @tparam _channel_dim  dimension in which channels are stored; channel dimension does not
+///                       participate in layout calculation \n
+///                       In cases where channel dimension can or should participate in layout
+///                       calaculation, do not specify channel dimenion and treat it as an
+///                       additional spatial dimension (e.g. for linear operations in CHW layout)\n
+///                       -1 indicates there are only spatial dimensions, all of which
+///                       participate in layout calculation.
+template <int _ndim, int _channel_dim>
+class BlockSetup {
+ public:
+  static constexpr int ndim = _ndim;
+  static constexpr int channel_dim = _channel_dim;
+  static constexpr int tensor_ndim = (channel_dim < 0 ? ndim : ndim + 1);
+  static_assert(channel_dim >= -1 && channel_dim <= ndim,
+    "Channel dimension must be in range [0..ndim] or -1 (no channel dim)");
+  using BlockDesc = kernels::BlockDesc<ndim>;
+
+  BlockSetup() {
+    for (int i = 0; i < ndim; i++)
+      default_block_size_[i] = (i < 2) ? 256 : 1;
+  }
+
+  void SetupBlocks(const TensorListShape<tensor_ndim> &output_shape,
+                   bool force_variable_size = false) {
+    blocks_.clear();
+    is_uniform_ = !force_variable_size && is_uniform(output_shape);
+
+    if (is_uniform_)
+      UniformSizeSetup(output_shape);
+    else
+      VariableSizeSetup(output_shape);
+  }
+
+  TensorListShape<tensor_ndim> GetOutputShape(
+      const TensorListShape<tensor_ndim> &in_shape,
+      span<const TensorShape<ndim>> output_sizes) {
+    TensorListShape<tensor_ndim> shape;
+    shape.resize(in_shape.num_samples(), tensor_ndim);
+    for (int i = 0; i < in_shape.num_samples(); i++) {
+      auto out_tshape = shape.tensor_shape_span(i);
+      int in_d = 0;
+      for (int j = 0; j < tensor_ndim; j++) {
+        out_tshape[j] = (j == channel_dim)
+         ? in_shape.tensor_shape_span(i)[channel_dim]
+         : output_sizes[i][in_d++];
+      }
+    }
+    return shape;
+  }
+
+
+  void ValidateOutputShape(
+      const TensorListShape<tensor_ndim> &out_shape,
+      const TensorListShape<tensor_ndim> &in_shape,
+      span<const TensorShape<ndim>> output_sizes) {
+    TensorListShape<tensor_ndim> shape;
+    shape.resize(in_shape.num_samples(), tensor_ndim);
```

Comment:
You can use just the constructor for this:
```suggestion
    TensorListShape<tensor_ndim> shape(in_shape.num_samples(), tensor_ndim);
```

Reply:
But is it ever used?

Reply:
**TADAM TADAM TADAM TADAM!**
That's what happens when you to make small PRs.

Reply:
Test added.